fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java

```diff
@@ -0,0 +1,249 @@
+import org.apache.doris.nereids.trees.plans.logical.LogicalSort;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalAggregate;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalDistribution;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalFilter;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalHashJoin;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalHeapSort;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalOlapScan;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalProject;
+import org.apache.doris.nereids.trees.plans.visitor.DefaultPlanVisitor;
+import org.apache.doris.nereids.util.Utils;
+import org.apache.doris.statistics.ColumnStats;
+import org.apache.doris.statistics.StatsDeriveResult;
+import org.apache.doris.statistics.TableStats;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Used to calculate the stats for each operator
+ */
+public class StatsCalculator extends DefaultPlanVisitor<StatsDeriveResult, Void> {
+
+    private final GroupExpression groupExpression;
+
+    public StatsCalculator(GroupExpression groupExpression) {
+        this.groupExpression = groupExpression;
+    }
+
+    /**
+     * Do estimate.
+     */
+    public void estimate() {
+        StatsDeriveResult stats = groupExpression.getPlan().accept(this, null);
+        groupExpression.getOwnerGroup().setStatistics(stats);
+        Plan plan = groupExpression.getPlan();
+        long limit = plan.getLimit();
+        if (limit != -1) {
+            stats.setRowCount(Math.min(limit, stats.getRowCount()));
+        }
+        groupExpression.setStatDerived(true);
+    }
+
+    @Override
+    public StatsDeriveResult visitLogicalAggregate(LogicalAggregate<Plan> agg, Void context) {
+        return computeAggregate(agg);
+    }
+
+    @Override
+    public StatsDeriveResult visitLogicalFilter(LogicalFilter<Plan> filter, Void context) {
+        return computeFilter(filter);
+    }
+
+    @Override
+    public StatsDeriveResult visitLogicalOlapScan(LogicalOlapScan olapScan, Void context) {
+        olapScan.getExpressions();
+        return computeScan(olapScan);
+    }
+
+    @Override
+    public StatsDeriveResult visitLogicalProject(LogicalProject<Plan> project, Void context) {
+        return computeProject(project);
+    }
+
+    @Override
+    public StatsDeriveResult visitLogicalSort(LogicalSort<Plan> sort, Void context) {
+        return groupExpression.getCopyOfChildStats(0);
+    }
+
+    @Override
+    public StatsDeriveResult visitLogicalJoin(LogicalJoin<Plan, Plan> join, Void context) {
+        return HashJoinEstimation.estimate(groupExpression.getCopyOfChildStats(0),
+                groupExpression.getCopyOfChildStats(1),
+                join.getCondition().get(), join.getJoinType());
```

Comment:
`HashJoinEstimation.estimate` need the input join condition is conjunctive equal-to predicates?

Reply:
Yes

Reply:
So we should filter equal-to condition here?

Reply:
It's still a hash join in the essence